core/src/block.rs

```diff
@@ -19,894 +19,635 @@ use iroha_data_model::{block::*, events::prelude::*, transaction::prelude::*};
-        ///
-        /// # Errors
-        /// Forward errors from [`CandidateBlock::revalidate`]
-        #[inline]
-        pub fn revalidate<const IS_GENESIS: bool>(
-            self,
-            transaction_validator: &TransactionValidator,
-            wsv: &WorldStateView,
-            latest_block: Option<HashOf<VersionedCommittedBlock>>,
-            block_height: u64,
-        ) -> Result<SignedBlock, eyre::Report> {
-            self.into_v1().revalidate::<IS_GENESIS>(
-                transaction_validator,
-                wsv,
-                latest_block,
-                block_height,
-            )
-        }
+    /// Check if a block has transactions that are already in the blockchain.
+    fn has_committed_transactions(&self, wsv: &WorldStateView) -> bool {
+        self.transactions
+            .iter()
+            .any(|transaction| transaction.is_in_blockchain(wsv))
+            || self
+                .rejected_transactions
+                .iter()
+                .any(|transaction| transaction.is_in_blockchain(wsv))
     }
+}
 
-    impl CandidateBlock {
-        /// Calculate the hash of the current block.
-
-        #[inline]
-        pub fn hash(&self) -> HashOf<Self> {
-            HashOf::new(&self.header).transmute()
-        }
-
-        /// Return signatures that are verified with the `hash` of this block, removing all other signatures.
-
-        #[inline]
-        pub fn retain_verified_signatures(&mut self) -> impl Iterator<Item = &SignatureOf<Self>> {
-            self.signatures.retain_verified_by_hash(self.hash())
-        }
-
-        /// Check if there are no transactions in this block.
-        #[inline]
-        pub fn is_empty(&self) -> bool {
-            self.transactions.is_empty() && self.rejected_transactions.is_empty()
-        }
-
-        /// Revalidate a block against the current state of the world.
-        ///
-        /// # Errors
-        /// - Block is empty
-        /// - Block has committed transactions
-        /// - There is a mismatch between candidate block height and actual blockchain height
-        /// - There is a mismatch between candidate block previous block hash and actual latest block hash
-        /// - Block header transaction hashes don't match with computed transaction hashes
-        /// - Error during revalidation of individual transactions
-        #[allow(clippy::too_many_lines)]
-        pub fn revalidate<const IS_GENESIS: bool>(
-            self,
-            transaction_validator: &TransactionValidator,
-            wsv: &WorldStateView,
-            latest_block: Option<HashOf<VersionedCommittedBlock>>,
-            block_height: u64,
-        ) -> Result<SignedBlock, eyre::Report> {
-            if self.is_empty() {
-                bail!("Block is empty");
-            }
-
-            if self.has_committed_transactions(wsv) {
-                bail!("Block has committed transactions");
-            }
+impl Revalidate for VersionedCommittedBlock {
```

Comment:
I'm a bit on the fence about implementing this for `VersionedCommittedBlock` instead of just `CommittedBlock`. If we moved all the helper methods from VCB into CB, both methods in this trait could have a default impl, as the destructuring from `V1` seems to be the only difference. And also semantically it seems to make more sense, as I believe versioning is more of a data model concern, and we shouldn't be dealing with it in core.

Reply:
Normally I'm not for future proofing but in this case I think it is important. By having all the methods on VCB we cover for the introduction of other block versions that do things differently. Currently we assume that a v1 block will always be convertible to a v2 block. This is not the case. Imagine just the simple scenario that after a v2 block has been committed we do not want to allow any more v1 blocks. As I've written it we simply switch on the block type and can implement this logic. The previous model were we convert everything into the latest version and use that would not support this case.

Reply:
I see, that makes quite some sense. Just had a natural urge to completely get rid of code reduplication, but it's not always achievable in real life.